Thread local fallback weak bag

vasilmkd opened the pull request:
Somewhat addresses #2663.

vasilmkd committed:
Use thread local weak bags as a fallback instead of locking

vasilmkd committed:
Remove unused class

vasilmkd committed:
Add a mima exclusion for the deleted class

armanbilge commented:
I know benchmarks are incoming 😉 but does this have the same basic caveat as https://github.com/typelevel/cats-effect/pull/2508#issuecomment-962658483?
> `ThreadLocal` is very expensive, and often exceeds the cost of contention, which is why you don't see it used too often. It also can have some complex GC implications which have their own performance costs.

vasilmkd commented:
I wanted to update the original comment, but I'll answer here.

Benchmarks are not coming because I'm not sure what to measure exactly.

The code that this replaces is _literally_ a `ThreadLocalRandom`, which is, you guessed it, a `ThreadLocal`. And in #2663, it was the locking contention that showed up, instead of the thread local usage.

I would like to call @yanns into the conversation if they would be willing to test this change out, with their workflow and measurement in Mission Control. Thanks in advance.

vasilmkd committed:
External runtime benchmark

vasilmkd commented:
Managed to come up with a benchmark.

`series/3.3.x`:
```scala
Benchmark                        (size)   Mode  Cnt    Score   Error  Units
ThreadLocalBenchmark.contention    2000  thrpt   20  287.505 ± 3.640  ops/s
```

`This PR`:
```scala
Benchmark                        (size)   Mode  Cnt    Score    Error  Units
ThreadLocalBenchmark.contention    2000  thrpt   20  306.480 ± 22.891  ops/s
```

In reality, the improvement is a bit misleading due to the bigger margin of error, so it might be a toss up, but thread locals are not strictly slower I guess.

vasilmkd commented:
Another run is more or less the same:

`This PR`:
```scala
Benchmark                        (size)   Mode  Cnt    Score    Error  Units
ThreadLocalBenchmark.contention    2000  thrpt   20  312.366 ± 30.404  ops/s
```

vasilmkd committed:
Explicit types are needed for implicit values

durban commented:
It's unclear to me, what guarantees that the `bag.toSet` call in `def foreignFibers()` "sees" the fibers inserted by `monitorFallback`, as `monitorFallback` accesses a bag directly, without any synchronization (and `WeakBag` itself does not seem thread-safe).

vasilmkd committed:
Add a synchronization point on insert

vasilmkd commented:
That's great input @durban. Thank you.

Another benchmark run with the latest changes:
`This PR`:
```scala
Benchmark                        (size)   Mode  Cnt    Score    Error  Units
ThreadLocalBenchmark.contention    2000  thrpt   20  327.031 ± 20.432  ops/s
```

vasilmkd committed:
Guard against memory leaks in strange executors

- If certain thread pools or executors cycle their threads, keeping weak
references to each bag lets those bags be eligible for GC when their
associated thread exits

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/FiberMonitor.scala:
```diff
@@ -172,4 +168,15 @@ private[effect] object FiberMonitor {
+  private[FiberMonitor] final val Bags: ThreadLocal[WeakBag[IOFiber[_]]] =
+    ThreadLocal.withInitial { () =>
+      val bag = new WeakBag[IOFiber[_]]()
+      BagReferences.offer(new WeakReference(bag))
```
@vasilmkd sorry, I had a follow-up question about this change.

Is it possible that we could lose track of suspended fibers, if the threads that they were suspended from no longer exist? Is that even a realistic situation 😆

vasilmkd replied:
That's a possibility, yes. The change was made with the intention that having an already inaccurate reporting mechanism remain that way is better than a memory leak. If people disagree, PRs are welcome.

armanbilge replied:
That's fair, thanks.

Probably over-complicated but I wonder if we could use a `PhantomReference` to "evacuate" the contents of the bag when its owning thread gets GCed.

Btw, since the WSTP also dynamically adds/removes threads, how is this problem handled there?

vasilmkd replied:
The WSTP does not use this code path. I'm open to exploring Phantom References.

armanbilge replied:
> The WSTP does not use this code path.

Right :) but it still uses a thread-local fiber bag right? And the threads may be added/removed as the WSTP resizes itself? So it seems like it's a very similar problem.

vasilmkd replied:
In the WSTP, each worker thread also has a thread local weak bag, in the sense that each thread has a unique reference to a bag. There is a transfer process of all important data structures when a thread is swapped out which solves this, which essentially ties the lifecycle of the bag to the pool, not to the thread, something which is not available on threads we do not control.

vasilmkd replied:
With Phantom References, there needs to be a dedicated thread which serves the reference queue when these bags expire, which polls references from it and does the cleanup. I'm not very enthusiastic about that but I can be persuaded.

armanbilge replied:
Yeah, I wasn't sure if it's worth it :) instead of a dedicated thread, is this something we can schedule on the runtime itself?

vasilmkd replied:
That's what we had before. It requires solving mapping of threads to bags, which was done using locking. If we come up with a concurrent weak bag/hash map, then sure. But not even JCTools has that afaik. It's a big undertaking.

Edit: I misunderstood your comment and answered something completely different.

vasilmkd replied:
Scheduling on the runtime requires answering how often do to run it, which to me doesn't seem like a good strategy for something considered to be memory beneficial/critical. And ReferenceQueue is not too smart of an interface either. You can poll it in a non-blocking way, and when it returns null, when do you try again? The proper way IMO is to block on it and run cleanup on each expiry.

armanbilge replied:
No it doesn't seem very elegant :) I feel like in practice, there must be some reasonable rate at which we can check the `ReferenceQueue` ... if an application is adding/removing threads too fast seems like its performance would be bounded by other factors anyway. But I don't really know about such things :)

After thinking about this more, seems like it could be important. A situation in which there is a deadlock seems like exactly the situation when a dynamically resizing threadpool would start culling threads due to lack of work, which could cause GC of the fiber bag holding the fibers would help diagnose the deadlock.

vasilmkd replied:
@djspiewak 👆🏻
